src/hotspot/share/classfile/javaClasses.cpp

```diff
@@ -2735,20 +2735,7 @@ void java_lang_Throwable::get_stack_trace_elements(int depth, Handle backtrace,
   }
 }
 
-Handle java_lang_Throwable::get_cause_with_stack_trace(Handle throwable, TRAPS) {
-  // Call to JVM to fill in the stack trace and clear declaringClassObject to
-  // not keep classes alive in the stack trace.
-  // call this:  public StackTraceElement[] getStackTrace()
-  assert(throwable.not_null(), "shouldn't be");
-
-  JavaValue result(T_ARRAY);
-  JavaCalls::call_virtual(&result, throwable,
-                          vmClasses::Throwable_klass(),
-                          vmSymbols::getStackTrace_name(),
-                          vmSymbols::getStackTrace_signature(),
-                          CHECK_NH);
-  Handle stack_trace(THREAD, result.get_oop());
-  assert(stack_trace->is_objArray(), "Should be an array");
+Handle java_lang_Throwable::get_cause(Handle throwable, bool with_stack_trace, TRAPS) {
```

Comment:
I see, you just reversed the code so it creates the message first, then calls getStackTrace.
If you change it to not have TRAPS, and use JavaThread* current as the first parameter, then have as the first line to show that the exception isn't propagated out of this function.
JavaThread* THREAD = current;
create the message
{ call getStackTrace();
  if (PENDING_EXCEPTION) CLEAR_PENDING_EXCEPTION;
  else set_stacktrace(); clear_backtrace();
}
return h_cause;

Something like that.  Then the caller doesn't need to check for null or pending exception.

Reply:
I agree with Coleen's suggestion, it doesn't make sense to return null and have a pending exception when the pending exception will be cleared in the caller anyway. This is a pre-existing issue but while we are here.